packages/providers/src/provider.ts

```diff
@@ -240,6 +289,17 @@ export default class Provider {
     return getOperationsSdk(gqlClient);
```

Comment:
Could you have done the caching on the level of the request by adding `requestMiddleware` to `GraphQLClient`? `requestMiddleware` gets called on every request the client makes and it seems like the ideal place to insert caching.

```ts
  private createOperations(url: string, options: ProviderOptions = {}) {
    this.url = url;

    const gqlClient = new GraphQLClient(url, {
      requestMiddleware: async (request) => {
        if (!this.consensusParamsCache[url]) {
          const consensusParameters = await Provider.getConsensusParamsWithoutInstance(url);
          this.consensusParamsCache[url] = consensusParameters;
        }
        return request;
      },
      fetch: options.fetch,
    });
    return getOperationsSdk(gqlClient);
  }
```

This way there wouldn't be any breaking changes to the public API of the `Provider` class as far as I can see, nor would there be changes in how to instantiate it.
The caching would happen on first **real** request instead of during instantiation of the `Provider`, which kind of makes sense because why would the provider fetch and cache consensus parameters for a node if it won't be contacting the node to do some actual querying/mutations/subscriptions?
Also, `createOperations` is called whenever `url` is changed, so we'd be sure that the first call to the new `url` would update the cache as well.


If somebody is actually interested in the consensus parameters, it can be put behind a method:

```ts
class Provider {
...
#consensusParametersCache: Record<string, ConsensusParameters | undefined>;

public async getConsensusParameters(): Promise<ConsensusParameters> {
    const cached = this.consensusParametersCache[this.url];
    if(cached) return cached;
    const consensusParameters = await Provider.getConsensusParamsWithoutInstance(this.url);
    this.consensusParametersCache[this.url] = consensusParameters;
    return consensusParameters;
  }
}
```

Reply:
This is very interesting @nedsalk. I'm intrigued. I'm gonna have a go at implementing this. Tagging the rest of the team so that they can have a look and share their opinions too cc @FuelLabs/sdk-ts

Reply:
@nedsalk I have an update on this. Like I mentioned in the weekly sync on Monday, this approach is really smart but will not work for us as of now. This is due to the fact that we need the consensus parameters when we _prepare_ a transaction and not only when we _send_ it. We only call the GQL API to _send_ the transaction. Hence, with the approach that you suggested, we won't be able to fetch the consensus parameters in time for things to work. Maybe we can make it work with a huge refactor and a re-design of the whole architecture (this would be possible in the Rust SDK, for example), but that would be outside the scope of this PR. If you want, we can still track this idea in this issue - #1103  to revisit later on.

Reply:
Thanks for the investigation and the explanation!

Given that this PR changes the way we create a `Provider` seemingly just so that we can satisfy the design that we find problematic, I think that we should put it on hold and focus on fixing that transaction design. If this PR was just introducing an internal change, I wouldn't raise this issue; however, this is introducing a breaking change but its purpose is lost because the consensus parameters we're looking to cache in the end aren't even used due to us fixating the values on transaction creation. 
The consensus parameters passed to a transaction on creation (be they cached or hardcoded) can become different to the latest ones the `Provider` has because X amount of time can pass between transaction creation and sending - and there is no way to change them as they're set _on creation_. When a transaction can be given consensus parameters the moment before it's sent instead of on its creation (e.g. by modifying `Provider.sendTransaction` to pass the consensus parameters to it and calculate the offsets at that moment), then it would make sense to revisit this issue and PR. Also, this PR is concerned with caching - meaning that everything works even without it but is just slower - so I don't see a big issue in postponing it.

@FuelLabs/sdk-ts please chime in.

Reply:
@nedsalk Although I completely see your point and I used to share this opinion a few weeks ago, at the moment I disagree mate. I think it will slow us down *a lot*. A massive architectural/design change like that will have to be extremely well thought-out. We need to ensure nothing breaks. We need this fix to go out for beta-4 and we don't have enough time to carefully craft a complete re-design.

Also, the architecture that this PR follows actually brings us closer to the Rust SDK. Which is why I believe this PR still takes us in a overall positive direction.

Reply:
Sounds good then!